src/RingGrids/general.jl

```diff
@@ -0,0 +1,398 @@
+"""Abstract supertype for all arrays of ring grids, representing `N`-dimensional
+data on the sphere in two dimensions (but unravelled into a vector in the first dimension,
+the actual "ring grid") plus additional `N-1` dimensions for the vertical and/or time etc.
+Parameter `T` is the `eltype` of the underlying data, held as in the array type `ArrayType`
+(Julia's `Array` for CPU or others for GPU).
+
+Ring grids have several consecuitive grid points share the same latitude (= a ring),
+grid points on a given ring are equidistant. Grid points are ordered 0 to 360˚E,
+starting around the north pole, ring by ring to the south pole. """
+abstract type AbstractGridArray{T, N, ArrayType <: AbstractArray{T, N}} <: AbstractArray{T, N} end
+
+"""Abstract supertype for all ring grids, representing 2-dimensional data on the
+sphere unravelled into a Julia `Vector`. Subtype of `AbstractGridArray` with
+`N=1` and `ArrayType=Vector{T}` of `eltype T`.""" 
+const AbstractGrid{T} = AbstractGridArray{T, 1, Vector{T}}
+
+## TYPES
+"""$(TYPEDSIGNATURES) For any instance of `AbstractGridArray` type its n-dimensional type
+(*Grid{T, N, ...} returns *Array) but without any parameters `{T, N, ArrayType}`"""
+nonparametric_type(grid::AbstractGridArray) = nonparametric_type(typeof(grid))
+
+"""$(TYPEDSIGNATURES) Full grid array type for `grid`. Always returns the N-dimensional `*Array`
+not the two-dimensional (`N=1`) `*Grid`. For reduced grids the corresponding full grid that
+share the same latitudes."""
+full_array_type(grid::AbstractGridArray) = full_array_type(typeof(grid))
+
+"""$(TYPEDSIGNATURES) Full (horizontal) grid type for `grid`. Always returns the two-dimensional
+(`N=1`) `*Grid` type. For reduced grids the corresponding full grid that share the same latitudes."""
+full_grid_type(grid::AbstractGridArray) = horizontal_grid_type(full_array_type(grid))
+full_grid_type(Grid::Type{<:AbstractGridArray}) = horizontal_grid_type(full_array_type(Grid))
+
+"""$(TYPEDSIGNATURES) The two-dimensional (`N=1`) `*Grid` for `grid`, which can be an N-dimensional
+`*GridArray`."""
+horizontal_grid_type(grid::AbstractGridArray) = horizontal_grid_type(typeof(grid))
+
+## SIZE
+Base.length(G::AbstractGridArray) = length(G.data)  # total number of grid points
+Base.size(G::AbstractGridArray) = size(G.data)      # size of underlying data (horizontal is unravelled)
+
+"""$(TYPEDSIGNATURES) Size of underlying data array plus precomputed ring indices."""
+Base.sizeof(G::AbstractGridArray) = sizeof(G.data) + sizeof(G.rings)
+
+"""$(TYPEDSIGNATURES) Resolution paraemeters `nlat_half` of a `grid`.
```

Comment:
typo

Reply:
You don't know parämeters? They're like parameters but crääzy!